### PR TITLE
Avoid insecure openssl_random_pseudo_bytes

### LINKED ISSUE #97: openssl_random_pseudo_bytes is insecure
It appears that openssl_random_pseudo_bytes is, in a large number of configurations, not safe for cryptographic use: https://github.com/paragonie/random_compat/issues/96 .

It might be better to switch to a PHP7 polyfill: https://github.com/paragonie/random_compat

### PR DESCRIPTION
**Pull request by skymeyer**
Fixes #97

### CODE REVIEW

File: phpunit.xml.dist [hunk 1]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<phpunit bootstrap="tests/bootstrap.php" colors="true">
+    <testsuites>
+        <testsuite name="xmlseclibs">
+            <directory suffix=".phpt">tests/</directory>
+        </testsuite>
+    </testsuites>
+    <filter>
+        <whitelist processUncoveredFilesFromWhitelist="true">
+            <directory suffix=".php">src/</directory>
```
Comment (skymeyer) [inline phpunit.xml.dist]: Coverage doesn't seem to work for me though, probably because of the usage of `phpt` test cases.

File: xmlseclibs.php [hunk 2]
```diff
@@ -40,6 +40,10 @@
  * @version   2.1.0-dev
  */
 
+if (!function_exists('random_bytes')) {
+    die('xmlseclibs requires https://github.com/paragonie/random_compat for PHP <7.0');
+}
```
Comment (skymeyer) [inline xmlseclibs.php]: Keeping this file for now. I'd argue to get rid of it entirely though with a proper version bump.

**Comment by relaxnow** [comment]
@robrichards could you take a look at this PR? It fixes a security vulnerability on configurations where PHP is forked.

**Comment by robrichards** [comment]
@relaxnow I'm looking at alternatives first as I would prefer not to be increasing external  dependancies unless absolutely necessary.

**Comment by relaxnow** [comment]
Actually this is a case where an external dependency makes a lot of sense. 

This is exactly why software like SimpleSAMLphp choses not to do it's own XML security. Or why SSP and OpenConext collaborate on a SAML2 package.

As in before this vulnerability was known the random_compat also preferred ```openssl_random_pseudo_bytes```. However when it was discovered in a package doing UUID generation, it was fixed in **random_compat** and would benefit all it's consumers.

Also this polyfill does nothing on PHP 7.

**Comment by robrichards** [comment]
Right now I'm waiting to find out what versions of PHP the fix for the openssl_random_pseudo_bytes function is going to be supported.

**Comment by relaxnow** [comment]
Looks like the patch is in:

>= 5.4.44
>= 5.5.28
>= 5.6.13
>= 7.0.0

Source: https://github.com/ramsey/uuid/issues/80#issuecomment-172002324

Also looks like this wasn't given a CVE so this wasn't ported by Red Hat: 
ftp://www.rpmfind.net/linux/RPM/centos/updates/6.7/x86_64/Packages/php-5.3.3-46.el6_6.x86_64.html

Still, seeing as [xmlseclibs supports PHP 5.3](https://github.com/robrichards/xmlseclibs/blob/master/composer.json#L18) this means that laggards like us are still suffering from this issue.

**Comment by robrichards** [comment]
That isn't the fix for this issue. There is currently one under review.

**Comment by skymeyer** [comment]
@robrichards any other alternatives to move forward on this issue ? This needs to be addressed.

**Comment by willemstuursma** [comment]
I would highly appreciate this change too. @robrichards, how to move forward?

File: src/XMLSecurityKey.php [hunk 3]
```diff
@@ -375,7 +368,7 @@ private function encryptMcrypt($data)
         if ($this->cryptParams['mode'] == MCRYPT_MODE_CBC) {
             $bs = mcrypt_enc_get_block_size($td);
             for ($datalen0 = $datalen = strlen($data); (($datalen % $bs) != ($bs - 1)); $datalen++)
```
Comment (lt) [inline src/XMLSecurityKey.php]: Looping over `random_int` for character codes is a little overkill. Not 100% sure what's going on here, looks like adding randomised padding? Do you know the reason for the selected character set? (Why hasn't a standard padding mode been used?)

File: src/XMLSecurityDSig.php [hunk 4]
```diff
@@ -152,7 +152,7 @@ private function getXPathObj()
      */
     public static function generateGUID($prefix='pfx')
     {
-        $uuid = md5(uniqid(mt_rand(), true));
+        $uuid = sha1(random_bytes(32));
```
Comment (skymeyer) [inline src/XMLSecurityDSig.php]: That makes more sense yes, will updated (already done in https://github.com/robrichards/xmlseclibs/pull/101)

File: src/XMLSecurityKey.php (hunk 3 above)
Comment (skymeyer) [inline src/XMLSecurityKey.php]: This random padding doesn't have any secure significance, leaving it untouched using `mt_rand`.

**Comment by skymeyer** [comment]
@robrichards I have retargeted this PR to the 2.0 release train and changed the code to the bare minimum. See https://github.com/robrichards/xmlseclibs/pull/101#issuecomment-267154917 for our options addressing the insecure `openssl_random_pseudo_bytes` issue.

Both this PR and https://github.com/robrichards/xmlseclibs/pull/101 can perfectly coexist. From a migration perspective I think it is better to have 3.0 release removing `mcrypt` with support for PHP 5.6+ and PHP 7.0+ instead of going PHP 7.0+ exclusively at this point.

**Comment by willemstuursma** [comment]
@robrichards how can we move this PR forward? Would love this released.

**Comment by robrichards** [comment]
As this only affects versions prior to 5.6.13 I personally don't see it as very critical. If however you really are looking to allow alternatives you would need to add optional support for it.

**Pull request closed without merging**